android/src/main/java/com/oblador/keychain/KeychainModule.java

```diff
@@ -37,99 +41,108 @@ public String getName() {
+            promise.reject(E_CRYPTO_FAILED, e);
         }
     }
 
     @ReactMethod
     public void getGenericPasswordForOptions(String service, Promise promise) {
-        service = service == null ? EMPTY_STRING : service;
-
-        String username = prefs.getString(service + ":u", "user_not_found");
-        String password = prefs.getString(service + ":p", "pass_not_found");
-        if (username.equals("user_not_found") || password.equals("pass_not_found")) {
-            Log.e(KEYCHAIN_MODULE, "no keychain entry found for service: " + service);
-            promise.resolve(false);
-            return;
-        }
-
-        byte[] recuser = Base64.decode(username, Base64.DEFAULT);
-        byte[] recpass = Base64.decode(password, Base64.DEFAULT);
-
-        Entity userentity = Entity.create(KEYCHAIN_DATA + ":" + service + "user");
-        Entity pwentity = Entity.create(KEYCHAIN_DATA + ":" + service + "pass");
-
         try {
-            byte[] decryptedUsername = crypto.decrypt(recuser, userentity);
-            byte[] decryptedPass = crypto.decrypt(recpass, pwentity);
+            CipherStorage currentCipherStorage = getCipherStorageForCurrentAPILevel();
+            if (currentCipherStorage == null) {
+                throw new CryptoFailedException("Unsupported Android SDK " + Build.VERSION.SDK_INT);
+            }
+
+            final DecryptionResult decryptionResult;
+            ResultSet resultSet = prefsStorage.getEncryptedEntry(service);
+            if (resultSet == null) {
+                Log.e(KEYCHAIN_MODULE, "No entry found for service: " + service);
+                promise.resolve(false);
+                return;
+            }
+
+            if (resultSet.cipherStorageName.equals(currentCipherStorage.getCipherStorageName())) {
+                // The encrypted data is encrypted using the current CipherStorage, so we just decrypt and return
+                decryptionResult = currentCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
+            }
+            else {
+                // The encrypted data is encrypted using an older CipherStorage, so we need to decrypt the data first, then encrypt it using the current CipherStorage, then store it again and return
+                CipherStorage oldCipherStorage = getCipherStorageByName(resultSet.cipherStorageName);
+                // decrypt using the older cipher storage
+                decryptionResult = oldCipherStorage.decrypt(service, resultSet.usernameBytes, resultSet.passwordBytes);
+                // encrypt using the current cipher storage
+                EncryptionResult encryptionResult = currentCipherStorage.encrypt(service, decryptionResult.username, decryptionResult.password);
+                // store the encryption result
+                prefsStorage.storeEncryptedEntry(service, encryptionResult);
+                // clean up the old cipher storage
+                oldCipherStorage.removeKey(service);
```

Comment:
I did not think that it was necessary to call `prefsStorage.resetPassword()` since the call to `prefsStorage.storeEncryptedEntry()` would overwrite any existing values. Can you see a scenario were it would not overwrite?

Also, I have the call to `oldCipherStorage.removeKey()` after `prefsStorage.storeEncryptedEntry()` in case the storing fails. If it does fail, then the old key is not removed.